MathLib/LinAlg/Lis/LisTools.cpp

```diff
@@ -14,34 +14,86 @@
 
 #include "LisTools.h"
 
+#include <cassert>
+
 #include "logog/include/logog.hpp"
 
 #include "LisMatrix.h"
 #include "LisVector.h"
 
+#include "BaseLib/quicksort.h"
+#include "MathLib/LinAlg/Sparse/CRSMatrix.h"
+#include "MathLib/LinAlg/Sparse/CRSTools.h"
+
 namespace MathLib
 {
 
-void applyKnownSolution(LisMatrix &A, LisVector &b, const std::vector<std::size_t> &vec_knownX_id,
-		const std::vector<double> &vec_knownX_x, double penalty_scaling)
+namespace detail
 {
-    //Use penalty parameter
-    const double max_diag_coeff = A.getMaxDiagCoeff();
-    const double penalty = max_diag_coeff * penalty_scaling;
-    INFO("-> max. absolute value of diagonal entries = %e", max_diag_coeff);
-    INFO("-> penalty scaling = %e", penalty_scaling);
-    const std::size_t n_bc = vec_knownX_id.size();
-    for (std::size_t i_bc=0; i_bc<n_bc; i_bc++) {
-        const std::size_t rowId = vec_knownX_id[i_bc];
-        const double x = vec_knownX_x[i_bc];
-        //A(k, k) = penalty
-        A.setValue(rowId, rowId, penalty);
-        //b(k) = x*penalty
-        b.set(rowId, x*penalty);
-    }
+MathLib::CRSMatrix<double, typename LisMatrix::IndexType>* lis2crs(LisMatrix &a)
```

Comment:
Some comments for the main steps of the algorithm would be helpful. (Yes, it's just a deep copy but maybe there is some special treatment like sorting or indices counting from 1...)